mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,247 @@
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+__doc__ = """
+.. autoclass:: StateConsumer
+.. autoclass:: PhysicalQuantity
+.. autoclass:: ConservedQuantity
+.. autoclass:: DependentQuantity
+.. autoclass:: KernelProfile
+.. autofunction:: add_package_versions
+.. autofunction:: set_state
+"""
+
+from logpyle import LogQuantity, LogManager
+from numpy import ndarray
+from mirgecom.profiling import PyOpenCLProfilingArrayContext
+from meshmode.discretization import Discretization
+from mirgecom.eos import GasEOS
+from meshmode.dof_array import DOFArray
+
+
+def add_package_versions(mgr: LogManager, path_to_version_sh: str = None) -> None:
+    """Add the output of the emirge version.sh script to the log."""
+    import os
+
+    if path_to_version_sh is None:
+        import pathlib
+        p = pathlib.Path(".").resolve()
+        for d in p.parents:
+            candidate = pathlib.Path(d).joinpath("version.sh")
+            if candidate.is_file() and "emirge" in candidate.open().read():
+                path_to_version_sh = str(candidate)
+                break
+
+    if path_to_version_sh is None:
+        from warnings import warn
+        warn("Could not find emirge's version.sh. No package versions recorded.")
+        mgr.set_constant("emirge_package_versions",
+            "Could not find emirge's version.sh. No package versions recorded.")
+
+    output = os.popen(path_to_version_sh).read()
+    if output == "":
+        output = "Could not record emirge's package versions."
+
+    mgr.set_constant("emirge_package_versions", output)
+
+
+def set_state(mgr: LogManager, state: ndarray) -> None:
+    """Update the state of all :class:`StateConsumer` of the log manager `mgr`."""
+    for gd_lst in [mgr.before_gather_descriptors,
+            mgr.after_gather_descriptors]:
+        for gd in gd_lst:
+            if isinstance(gd.quantity, StateConsumer):
+                gd.quantity.set_state(state)
+
+
+class StateConsumer:
+    """Base class for quantities that require a state for logging."""
+
+    def __init__(self):
+        self.state = None
+
+    def set_state(self, state: ndarray) -> None:
+        """Set the state of the object."""
+        self.state = state
+
+
+class PhysicalQuantity(LogQuantity, StateConsumer):
+    """Logging support for physical quantities."""
+
+    def __init__(self, discr: Discretization, quantity: str, unit: str, op: str,
+                 name: str, rank_aggr: str = None):
```

Comment:
Why is `rank_aggr` a constructor parameter? (When it only sets the default, which can be overridden by the user when querying?)

Reply:
Removed.